src/gamestate/system_state.cpp

```diff
@@ -4539,385 +4144,962 @@ void state::console_log(std::string_view message) {
 }
 
 void state::new_army_group(dcon::province_id hq) {
-	for(auto& item : army_groups) {
-		if(item.hq == hq) {
-			return;
+	bool invalid_province = false;
+	world.for_each_automated_army_group(
+		[&](dcon::automated_army_group_id item) {
+			auto item_hq = world.automated_army_group_get_hq(item);
+			if(item_hq == hq) {
+				invalid_province = true;
+			}
 		}
+	);
+	if(invalid_province) {
+		return;
 	}
-
-	army_group new_group = {
-		.hq = hq,
-		.land_forces = { },
-		.land_regiments = { },
-		.naval_forces = { },
-		.ships = { },
-
-		.defensive_line = { },
-		.naval_travel_origin = { },
-		.naval_travel_target = { }
-	};
-
-	army_groups.push_back(new_group);
+	auto new_group = world.create_automated_army_group();
+	world.automated_army_group_set_hq(new_group, hq);
+	world.automated_army_group_set_owner(new_group, local_player_nation);
 
 	game_state_updated.store(true, std::memory_order_release);
 }
 
-void state::toggle_defensive_position(army_group* group, dcon::province_id position) {
-	auto index = std::find(group->defensive_line.begin(), group->defensive_line.end(), position);
-	if(index != group->defensive_line.end()) {
-		group->defensive_line.erase(index);
-		return;
+void state::toggle_defensive_position(dcon::automated_army_group_id group, dcon::province_id position) {
+	auto fat_group = fatten(world, group);
+
+	if(fat_group.get_provinces_defend().contains(position)) {
+		fat_group.get_provinces_defend().remove_unique(position);
+	} else {
+		fat_group.get_provinces_defend().add_unique(position);
```

Comment:
micro optimization: if you know that it isn't in the list, then just pushing it to the back is cheaper